src/UndockedRegFreeWinRT/UndockedRegFreeWinRT/typeresolution.cpp

```diff
@@ -0,0 +1,653 @@
+                                }
+
+                                if (hr == RO_E_METADATA_NAME_IS_NAMESPACE)
+                                {
+                                    break;
+                                }
+                            }
+                        } while (FindNextFile(hFindFile, &fd));
+
+                        FindClose(hFindFile);
+                    }
+                    else
+                    {
+                        hr = RO_E_METADATA_NAME_NOT_FOUND;
+                    }
+                }
+            }
+        }
+
+        if (hr == STRSAFE_E_INSUFFICIENT_BUFFER)
+        {
+            hr = RO_E_METADATA_NAME_NOT_FOUND;
+        }
+
+        return hr;
+    }
+
+    HRESULT FindTypeInDirectoryWithNormalization(
+        __in IMetaDataDispenserEx* pMetaDataDispenser,
+        __in PCWSTR pszFullName,
+        __in PCWSTR pszDirectoryPath,
+        __out_opt HSTRING* phstrMetaDataFilePath,
+        __deref_opt_out_opt IMetaDataImport2** ppMetaDataImport,
+        __out_opt mdTypeDef* pmdTypeDef)
+    {
+        wchar_t pszPackagePath[MAX_PATH + 1];
+        PWSTR pszPackagePathWritePtr = pszPackagePath;
+        size_t cchPackagePathRemaining = ARRAYSIZE(pszPackagePath);
+
+        HRESULT hr = StringCchCopyEx(
+            pszPackagePath,
+            ARRAYSIZE(pszPackagePath),
+            pszDirectoryPath,
+            &pszPackagePathWritePtr,
+            &cchPackagePathRemaining,
+            0);
+
+        if (SUCCEEDED(hr))
+        {
+            // If the path is not terminated by a backslash, then append one.
+            if (pszPackagePath[ARRAYSIZE(pszPackagePath) - cchPackagePathRemaining - 1] != L'\\')
+            {
+                hr = StringCchCopyEx(
+                    pszPackagePathWritePtr,
+                    cchPackagePathRemaining,
+                    L"\\",
+                    &pszPackagePathWritePtr,
+                    &cchPackagePathRemaining,
+                    0);
+            }
+        }
+
+        if (SUCCEEDED(hr))
+        {
+            hr = FindTypeInDirectory(
+                pMetaDataDispenser,
+                pszFullName,
+                pszPackagePath,
+                phstrMetaDataFilePath,
+                ppMetaDataImport,
+                pmdTypeDef);
+        }
+
+        return hr;
+    }
+
+    HRESULT ResolveThirdPartyType(
+        __in IMetaDataDispenserEx* pMetaDataDispenser,
+        __in PCWSTR pszFullName,
+        __out_opt HSTRING* phstrMetaDataFilePath,
+        __deref_opt_out_opt IMetaDataImport2** ppMetaDataImport,
+        __out_opt mdTypeDef* pmdTypeDef)
+    {
+        HRESULT hr = HRESULT_FROM_WIN32(APPMODEL_ERROR_NO_PACKAGE);
+        UINT32 dwPackagesCount = 0;
+        UINT32 dwBufferLength = 0;
+
+        PCWSTR exeDir = nullptr;  // Never freed; owned by process global.
+        RETURN_IF_FAILED(GetProcessExeDir(&exeDir));
+
+        hr = FindTypeInDirectoryWithNormalization(
```

Comment:
Would it make sense to restrict exe dir resolutions to only apply to unpackaged apps? That's how it's done in the OS-provided API, and I think consistency with that behavior should be our default approach. Packaged apps can have their exe on their package graph, which will allow type resolution against adjacent WinMDs.

Reply:
That depends on how it works down level to rs2 and windows 8.1.
https://github.com/microsoft/xlang/issues/687

Would the old os versions be able to handle it for the packaged apps?

Reply:
Yes, packaged type resolution was the original goal of these APIs, and its behavior has been mostly unchanged since Win8. Unpackaged type resolution was the new thing added with reg-free WinRT, so it's the only thing that you should need to undock and take downlevel via detours.

Reply:
Alright then it makes sense to add the check then.